### PR TITLE
fix: no more disappearing canister ids from .env files

### PR DESCRIPTION
**Pull request by ericswanson-dfinity**
# Description

Made it so `dfx deploy` and `dfx canister install` will always write canister ids for
all canisters in the project to the .env file, even if they aren't being deployed/installed or
a dependency of a canister being deployed/installed.

Fixes https://dfinity.atlassian.net/browse/SDK-1369

# How Has This Been Tested?

Updated an e2e test (dfx build) and added two more (dfx deploy, dfx canister install)

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

File: e2e/tests-dfx/build.bash [hunk 1]
```diff
@@ -96,6 +96,14 @@ teardown() {
   assert_match "syntax error"
 }
 
+@test "build --check fails on invalid motoko" {
+  install_asset invalid
+  dfx_start
+  # --check does not need create --all
```
Comment (lwshang) [inline e2e/tests-dfx/build.bash:102]:
```suggestion
```

`dfx build --check` needs not `dfx start`.
There are three other tests above also only run `dfx build` with `--check`.

File: e2e/tests-dfx/dotenv.bash [hunk 2]
```diff
@@ -51,15 +51,48 @@ teardown() {
...
+  dfx_start
+  dfx canister create --all
+  dfx build e2e_project_frontend
+  # set a post-install script so the install will create a .env file
+  jq '.canisters.e2e_project_frontend.post_install="echo post install"' dfx.json | sponge dfx.json
+
+  writes_all_environment_variables_to_env canister install
+}
+
+writes_all_environment_variables_to_env() {
```
Comment (lwshang) [inline e2e/tests-dfx/dotenv.bash:78]:
```suggestion
assert_dotenv_contains_all_variables_after_command() {
```

I was confused when I see no assertions in the new tests above.
Then I found out that this method make assertions inside.

**Review by lwshang — APPROVED** [review]
LGTM. Only two minor comments about the tests.

File: e2e/tests-dfx/build.bash (hunk 1 above)
Comment (ericswanson-dfinity) [reply e2e/tests-dfx/build.bash:102]: I'm going to add a test that still calls `dfx start` and `dfx canister create --all` and verifies `dfx build --check`.  The reason is that there are slightly different code paths involved.

An earlier commit on this PR broke `dfx build --check` when there are no canister ids, but not when canister ids are known.

**Pull request merged**